File: string_process.py
import re
import time
import datetime
from collections import defaultdict

def to_list(file):
  f = open(file, 'rt', encoding = 'UTF-8')
  lines = f.readlines()
  return lines

def to_dict(items):
  grouper = defaultdict(list)
  for item in items:
    item_sub = re.sub('[,.?! ~\n]','',item)
    grouper[len(item_sub)].append(item)
  return grouper

w_punct = to_list('broadcast_w_punct_u.txt')
w_punct_dict = to_dict(w_punct)
# print(w_punct_dict)

f = open('SubtTV_2017_01_03_pcm.list.trn', 'rt', encoding = 'UTF-8')
new_f = open('SubtTV_2017_01_03_pcm.list.punct.trn', 'wt', encoding = 'UTF-8')
not_found = open('SubtTV_2017_not_found', 'wt', encoding = 'UTF-8')

print(datetime.datetime.now())
while True:
  # counter = time.time()

  line = f.readline()
  if not line : break

  file_name = line.split(' :: ')[0]
  org_txt = line.split(' :: ')[1][:-1]
  
  org_txt_sub = re.sub('[ ]', '', org_txt)
  cmp_len = len(org_txt_sub)

  first_word = org_txt.split(' ')[0]
  second_word = org_txt.split(' ')[-1]
  # print('org : ', org_line_sub)
  flag = True
  while flag:
    cmp_list = w_punct_dict[cmp_len]
    if not cmp_list:
      # print('fail : ', time.time() - counter)
      not_found.write(line)
      flag = False
    for cmp_line in cmp_list:
      cmp_txt = cmp_line[:-1]
      cmp_txt_sub = re.sub('[,.?!~ ]', '', cmp_txt)
      # print('cmp : ', cmp_line_sub)
      if org_txt_sub in cmp_txt_sub:
        start = cmp_txt.find(first_word)
        check = cmp_txt.find(second_word, start + len(first_word))
        end = cmp_txt.find(' ', check)
        if end == -1:
          new_f.write(file_name + ' :: ' + cmp_txt[start:] + '\n')
        else:
          new_f.write(file_name + ' :: ' + cmp_txt[start:end] + '\n')
        # print('done : ', time.time() - counter)
        flag = False
        break
    cmp_len += 1

File: string_process.go
package main

import (
	"bufio"
	"os"
	"regexp"
	"strings"
)

func ToSlice(path string) []string {
	f, err := os.Open(path)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	// println(lines)
	return lines
}

func ToMap(lines []string) map[int][]string {
	var lenMap map[int][]string
	lenMap = make(map[int][]string)

	for _, line := range lines {
		var re = regexp.MustCompile("[,.?! ~\n]")
		lineSub := re.ReplaceAllString(line, "")
		// println(lineSub)
		_, exists := lenMap[len(lineSub)]
		if !exists {
			lenMap[len(lineSub)] = []string{line}
		} else {
			lenMap[len(lineSub)] = append(lenMap[len(lineSub)], line)
		}
	}
	return lenMap
}

func FileOpen(path string) *os.File {
	f, err := os.Open(path)
	if err != nil {
		panic(err)
	}

	defer func() {
		if err := f.Close(); err != nil {
			panic(err)
		}
	}()

	return f
}

func main() {
	wPunct := ToSlice(`C:\Project\20200804.-방송DB후처리\broadcast_w_punct_u.txt`)
	wPunctMap := ToMap(wPunct)
	f, err := os.Open(`C:\Project\20200804.-방송DB후처리\SubtTV_2017_01_03_pcm.list.trn`)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	newF, err := os.Create(`C:\Project\20200804.-방송DB후처리\SubtTV_2017_01_03_pcm.list.punct.trn`)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	notFound, err := os.Create(`C:\Project\20200804.-방송DB후처리\SubtTV_2017_not_found.trn`)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var line string
	scanner := bufio.NewScanner(f)

	for scanner.Scan() {
		line = scanner.Text()
		fileName := strings.Split(line, " :: ")[0]
		orgTxt := strings.Split(line, " :: ")[1]
		orgTxtSub := strings.ReplaceAll(orgTxt, " ", "")

		cmpLen := len(orgTxtSub)
		point := cmpLen + 5

		splitOrgTxt := strings.Split(orgTxt, " ")
		firstWord := splitOrgTxt[0]
		lastWord := splitOrgTxt[len(splitOrgTxt)-1]

		flag := true

		for flag {
			cmpList, exists := wPunctMap[cmpLen]
			if !exists || cmpLen > point {

				flag = false
				break
			}
			var re = regexp.MustCompile("[,.?! ~\n]")
			for _, cmpTxt := range cmpList {
				cmpTxtSub := re.ReplaceAllString(cmpTxt, "")
				if strings.Contains(cmpTxtSub, orgTxtSub) {
					start := strings.Index(cmpTxt, firstWord)
					check := strings.Index(cmpTxt[start+len(firstWord):], lastWord)
					check = check + start + len(firstWord)
					end := strings.Index(cmpTxt[check+len(lastWord):], " ")
					if end == -1 {

					} else {
						end = end + check + len(lastWord)
					}
					flag = false
					break
				}
			}
			cmpLen++
		}
	}

	// print(f, newF, notFound)
	// for key, _ := range wPunctMap {
	// 	println(key)
	// }
	// print(wPunctMap[53][0])
	// print(wPunctMap[40][0])
	// print(wPunctMap[30])
}
